internal/delivery/http/v1/auth/bearer_auth_integration_test.go

```diff
@@ -0,0 +1,211 @@
+package auth_test
+
+import (
+	"encoding/json"
+	"github.com/gin-gonic/gin"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/zhuravlev-pe/course-watch/internal/delivery/http/v1/auth"
+	"github.com/zhuravlev-pe/course-watch/pkg/security"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+)
+
+var validKey = []byte("1234")
+var invalidKey = []byte("42-42-42")
+
+func getParametrizedTokenHandler(key []byte) *security.JwtHandler {
+	jwt := security.NewJwtHandler()
+
+	const aud = "course-watch-api"
+
+	jwt.Issuer = "course-watch"
+	jwt.AudienceExpected = aud
+	jwt.AudienceGenerated = []string{aud}
+	jwt.SigningKey = key
+	jwt.TokenTtl = time.Hour * 1
+
+	return jwt
+}
+
+func TestBearerAuthenticator_Integration_Authorize(t *testing.T) {
+	tokenHandler := getParametrizedTokenHandler(validKey)
+	fakeTokenHandler := getParametrizedTokenHandler(invalidKey)
+
+	ba := auth.NewBearerAuthenticator(tokenHandler)
+
+	var endpointHit bool
+
+	router := gin.New()
+	g := router.Group("/secure", ba.Authorize(security.Admin))
```

Comment:
This shows the idea of the new middleware: we just put it on a gin router group and the endpoints will be hit only if the caller provides a valid bearer token in the `Authorization` header. Otherwise, middleware automatically generates 401 or 403
`Authorize()` requires the user to be authenticated and ensures that the user has the specified role. `Authenticate` checks only for a valid token. Function `GetAuthenticatedUser()` can be called in the endpoint body to access the logged in user's data (security.UserPrincipal, contains only Id and roles for now)

Reply:
Great!